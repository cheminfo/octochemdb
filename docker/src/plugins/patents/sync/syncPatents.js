import { existsSync, rmSync } from 'fs-extra';
import md5 from 'md5';

import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import debugLibrary from '../../../utils/Debug.js';

import firstPatentsImport from './utils/firstPatentsImport.js';
import ungzipAndSort from './utils/ungzipAndSort.js';

/**
 * @description sync patents from PubChem database
 * @param {*} connection - mongo connection
 * @returns {Promise} returns patents collection
 */
export async function sync(connection) {
  const debug = debugLibrary('syncPatents');
  try {
    let options = {
      collectionSource: process.env.CIDTOPATENT_SOURCE,
      destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/patents/cidToPatents`,
      collectionName: 'patents',
      filenameNew: 'cidToPatents',
      extensionNew: 'gz',
    };
    let sources;
    let lastFile;
    const progress = await connection.getProgress('patents');
    if (process.env.NODE_ENV === 'test') {
      lastFile = `${process.env.PATENTS_SOURCE_TEST}`;
      sources = [lastFile];
    }
    // get last files cidToPatens available in the PubChem database
    else if (
      Date.now() - Number(progress.dateEnd) >
      Number(process.env.PATENT_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000
    ) {
      lastFile = await getLastFileSync(options);
      sources = [lastFile.replace(`${process.env.ORIGINAL_DATA_PATH}`, '')];
    } else {
      sources = progress.sources; // this will prevent to update the collection
    }

    let shouldUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - Number(progress.dateEnd) >
        Number(process.env.PATENT_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000 &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      shouldUpdate = true;
      await connection.setProgress(progress);
    }
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    if (
      (JSON.stringify(sources) !== progress.sources && shouldUpdate) ||
      progress.state !== 'updated'
    ) {
      progress.state = 'updating';
      await connection.setProgress(progress);
      //sort file by cid
      const sortedFile = `${lastFile.split('.gz')[0]}.sorted`;
      await ungzipAndSort(lastFile, sortedFile);
      debug('ungzip and sort done');

      await firstPatentsImport(sortedFile, connection);
      const collection = await connection.getCollection(options.collectionName);
      await collection.createIndex({ 'data.patents': 1 });
      await collection.createIndex({ 'data.nbPatents': 1 });
      await collection.createIndex({ _seq: 1 });

      // update Logs in importationLogs collection
      progress.sources = md5(JSON.stringify(sources));
      progress.state = 'updated';
      progress.dateEnd = Date.now();
      await connection.setProgress(progress);

      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';

      await connection.updateImportationLog(logs);
      // remove recursively the sorted file
      if (existsSync(sortedFile)) {
        rmSync(sortedFile, { recursive: true });
      }
    }
  } catch (e) {
    if (connection) {
      debug(e.message, { collection: 'patents', connection, stack: e.stack });
    }
  }
}
