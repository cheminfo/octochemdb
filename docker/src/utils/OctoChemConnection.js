import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
