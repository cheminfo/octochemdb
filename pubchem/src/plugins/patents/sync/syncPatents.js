import pkg from 'fs-extra';
import md5 from 'md5';

import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import Debug from '../../../utils/Debug.js';

import firstPatentsImport from './utils/firstPatentsImport.js';
import ungzipAndSort from './utils/ungzipAndSort.js';

const { removeSync } = pkg;
/**
 * @description sync patents from PubChem database
 * @param {*} connection - mongo connection
 * @returns {Promise} returns patents collection
 */
export async function sync(connection) {
  const debug = Debug('syncPatents');
  try {
    let options = {
      collectionSource: process.env.CIDTOPATENTS_SOURCE,
      destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/patents/cidToPatents`,
      collectionName: 'patents',
      filenameNew: 'cidToPatents',
      extensionNew: 'gz',
    };

    // get last files cidToPatens available in the PubChem database
    const lastFile = await getLastFileSync(options);
    const sources = [lastFile.replace(process.env.ORIGINAL_DATA_PATH, '')];
    const progress = await connection.getProgress('patents');
    if (
      progress.dateEnd !== 0 &&
      progress.dateEnd - Date.now() > process.env.PATENT_DATE_INTERVAL &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      await connection.setProgress(progress);
    }
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    if (
      JSON.stringify(sources) !== progress.sources ||
      progress.state !== 'updated'
    ) {
      progress.state = 'updating';
      await connection.setProgress(progress);
      //sort file by cid
      const sortedFile = `${lastFile.split('.gz')[0]}.sorted`;
      await ungzipAndSort(lastFile, sortedFile);
      //  remove non-sorted file
      removeSync(lastFile);
      //  const sortedFile = `${options.destinationLocal}/cidToPatents.2022-07-14.sorted`;
      await firstPatentsImport(sortedFile, connection);
      const collection = await connection.getCollection(options.collectionName);
      await collection.createIndexes(
        { _id: 1 },
        { 'data.patents': 1 },
        { 'data.nbPatents': 1 },
      );

      // update Logs in importationLogs collection
      progress.sources = md5(JSON.stringify(sources));
      progress.state = 'updated';
      progress.dateEnd = Date.now();
      await connection.setProgress(progress);

      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';

      await connection.updateImportationLog(logs);
    }
  } catch (e) {
    if (connection) {
      debug(e.message, { collection: 'patents', connection, stack: e.stack });
    }
  }
}
