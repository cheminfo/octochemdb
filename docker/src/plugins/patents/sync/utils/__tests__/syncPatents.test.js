import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncPatents';

test(
  'syncPatents',
  async () => {
    const connection = new OctoChemConnection();
    await sync(connection);
    const collection = await connection.getCollection('patents');
    const collectionEntry = await collection
      .find({ _id: 'EP-2078065-A2' })
      .limit(1);
    const result = await collectionEntry.next();

    expect(result).toMatchSnapshot();
    await connection.close();
  },
  { timeout: 300000 },
);
