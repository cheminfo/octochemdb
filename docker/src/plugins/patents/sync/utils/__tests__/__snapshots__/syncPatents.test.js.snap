// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncPatents 1`] = `
{
  "_id": "EP-2078065-A2",
  "data": {
    "abstract": "This invention relates to a method of processing color photographic film that has been imagewise exposed in a camera, said film having at least three light-sensitive units which have their individual sensitivities in different wavelength regions, each of the units comprising at least one light-sensitive silver-halide emulsion, binder, and dye-providing coupler, which method in order comprises (a) thermally developing the film step without any externally applied developing agent, comprising heating said film to a temperature greater than 80 Â°C in an essentially dry process, such that an internally located blocked developing agent in reactive association with each of said three light-sensitive units becomes unblocked to form a developing agent, whereby the unblocked developing agent forms dyes by reacting with the dye-providing couplers to form a color image; (b) scanning the color image in the film without desilvering; (c) desilvering said film in one or more desilvering solutions to remove at least silver halide, thereby forming an improved color image suitable for scanning or optical printing, and (d) either optically printing or scanning the color image in the film following desilvering. In one embodiment of the invention, the film is scanned a first time in step (b) to obtain a relatively low quality scan and then scanned a second time after step (c) to obtain a relatively high quality scan that is used for making the positive image print.",
    "nbCompounds": 1,
    "title": "Process for preparing optically active epoxy compound",
  },
}
`;
